File: Assets/Scripts/JavaScript/InteractablesJoe/PickupGun.js
import UnityEngine.UI;

var distanceToObject: float = PlayerCasting.DistanceFromTarget;     // Distance from the Player to the game object
var TextDisplay: GameObject;                                        // Canvas action text

var FakeGun: GameObject;                                            // The collectable gun object
var RealGun: GameObject;                                            // The gun the player is carrying
var AmmoDisplay: GameObject;                                        // Shows the text with the amount of ammo the player has
var TargetDisplay: GameObject;										// Shows the text with the amount of targets the player has to shoot
var PickupAudio: AudioSource;                                       // Sound effect for picking up ammo

var hasGun: GameObject;                                             // Set the hasGun objective to true

function Update() {
    distanceToObject = PlayerCasting.DistanceFromTarget;             // Use raycasting to calculate the distance from the player to the object
}

function OnMouseOver() {                                            // If the cursor is over the object
    if (distanceToObject <= 2) {                                     // and the distance is less than 2
        TextDisplay.GetComponent.<Text>().text = "Take 9mm Pistol"; // Display the message
    }       
    if (Input.GetButtonDown("Action")) {                            // If the action button (E) is pressed
        if (distanceToObject <= 2) {                                 // and the distance is less than 2
            TakeGun();                                              // call the take gun function
        }
    }
}

function OnMouseExit() {                                            // When the cursor has moved away from the object
    TextDisplay.GetComponent.<Text>().text = "";                    // Reset the action text
}

function TakeGun() {
    hasGun.SetActive(true);                                         // Set the hasGun objective as true

    PickupAudio.Play();                                             // play the sound effect
    transform.position = Vector3(0, -1000, 0);                      // hide the weapon out of the scene
    FakeGun.SetActive(false);                                       // hide pickup gun
    RealGun.SetActive(true);                                        // display player gun
    AmmoDisplay.SetActive(true);                                    // Set the UI for displaying the Players ammo count active
    TargetDisplay.SetActive(true);                                  // Set the UI for displaying the targets to shoot count active

    // CANCELLED OUT BY OnMouseExit()
    //TextDisplay.GetComponent.<Text>().text = "Objective 1:\nFind Gun Complete"; // Display the message
    //yield WaitForSeconds(2);                            			// Wait one second (half way through the 2 second animation)
    //TextDisplay.GetComponent.<Text>().text = ""; 					// Display the message
}

File: Assets/Scripts/JoeCS/Weapons/PickUpGun.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;

public class PickUpGun : MonoBehaviour {

	public float distanceTo;

	Text infoMsg;

	public GameObject playerGun;
	//public GameObject pickupGun;
	public GameObject ammoText;
	public GameObject targetText;
	public GameObject hasGunObjective;

	private AudioSource pickupAudio;

	private GameObject mainCam;
	private GameObject gc;
	private GameObject player;

	// Use this for initialization
	void Start () {
		mainCam = GameObject.FindGameObjectWithTag ("MainCamera");
		gc = GameObject.FindGameObjectWithTag ("GameController");
		player = GameObject.FindWithTag("Player");
		pickupAudio = GetComponent<AudioSource> ();		
		infoMsg = GameObject.FindWithTag ("InfoMessage").GetComponent<Text> ();
	}
	
	// Update is called once per frame
	void Update () {
		distanceTo = mainCam.GetComponent<PlayerCasting2> ().getDistanceFromTarget ();			
	}

	void OnMouseOver() {
		//Debug.Log ("Mouse over Chainsaw");

		if (distanceTo < 2.0f) {
			infoMsg.text = "Take 9mm Pistol";
			StartCoroutine ("ClearText");							// Clear text after 2 seconds

			if (Input.GetButtonDown ("Action") && gc.GetComponent<ManageZombies> ().totalZombieKills () == 0) {  
				StartCoroutine(PickupGun ());						// Get the gun
			}
		}
	}

	IEnumerator ClearText(){
		yield return new WaitForSeconds (2);
		infoMsg.text = "";
	}

	IEnumerator PickupGun(){
		player.GetComponent<WeaponSelect> ().HasGun(true);			// Enable the gun for the player

		playerGun.gameObject.SetActive (true);						// Show the Player gun
		ammoText.gameObject.SetActive (true);						// Show the ammo count text on UI panel
		targetText.gameObject.SetActive (true);						// Show the targets to kill on UI panel
		hasGunObjective.gameObject.SetActive (true);				// Activate the hasGun objective object

		pickupAudio.Play();                                        	// play the sound effect

		yield return new WaitForSeconds (2);

		this.gameObject.SetActive (false);							// Hide the pickup gun object
	}
}
